onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc

```diff
@@ -5,73 +5,180 @@
+    bool simplified,
+    T* Y_data,
+    U* mean_data,
+    U* inv_std_dev_data) {
+  ORT_UNUSED_PARAMETER(scale_fp32);  // only used in MLFloat16 overload
+  ORT_UNUSED_PARAMETER(bias_fp32);   // only used in MLFloat16 overload
+
+  const T* p_input = X_data + task_idx * norm_size;
+  T* p_output = Y_data + task_idx * norm_size;
+
+  T mean(0.0f);
+  T mean_square(0.0f);
+
+  for (int64_t h = 0; h < norm_size; h++) {
+    p_output[h] = p_input[h];
+    mean += p_input[h];
+    mean_square += p_input[h] * p_input[h];
+  }
 
-template <>
-ORT_FORCEINLINE double ConvertMLFloat16ToDoubleOrFloatIfNeeded<MLFloat16, double>(MLFloat16 val) {
-  return double(ConvertMLFloat16ToDoubleOrFloatIfNeeded<MLFloat16, float>(val));
-}
+  mean = mean / norm_size;
+  if (simplified) {
+    mean_square = sqrt(mean_square / norm_size + epsilon);
+  } else {
+    mean_square = sqrt(mean_square / norm_size - mean * mean + epsilon);
+  }
 
-template <>
-ORT_FORCEINLINE constexpr float ConvertMLFloat16ToDoubleOrFloatIfNeeded<float, float>(float val) {
-  return val;
-}
+  for (int64_t h = 0; h < norm_size; h++) {
+    if (simplified) {
+      p_output[h] = p_output[h] / mean_square * scale_data[h];
+    } else if (nullptr == bias_data) {
+      p_output[h] = (p_output[h] - mean) / mean_square * scale_data[h];
+    } else {
+      p_output[h] = (p_output[h] - mean) / mean_square * scale_data[h] + bias_data[h];
+    }
+  }
 
-template <>
-ORT_FORCEINLINE constexpr double ConvertMLFloat16ToDoubleOrFloatIfNeeded<double, double>(double val) {
-  return val;
-}
+  if (mean_data != nullptr) {
+    // ONNX spec doesn't support 'double' for 'U' so when 'T' == double, 'U' == float and we need to narrow
+    mean_data[task_idx] = gsl::narrow_cast<float>(mean);
+  }
 
-ORT_FORCEINLINE constexpr float ConvertToFloatIfNeeded(float val) {
-  return val;
+  if (inv_std_dev_data != nullptr) {
+    inv_std_dev_data[task_idx] = gsl::narrow_cast<float>(1 / mean_square);
+  }
 }
 
-ORT_FORCEINLINE constexpr float ConvertToFloatIfNeeded(double val) {
-  // ONNX spec doesn't support 'double' for 'Ret' so when 'T' == double, 'Ret' == float and we need to narrow
-  return gsl::narrow_cast<float>(val);
-}
+template <typename U>
+void ComputeJob(
+    const MLFloat16* X_data,
+    const MLFloat16* scale_data,
+    const MLFloat16* bias_data,
+    const ptrdiff_t task_idx,
+    const int64_t norm_size,
+    const IAllocatorUniquePtr<float>& scale_fp32,
+    const IAllocatorUniquePtr<float>& bias_fp32,
+    float epsilon,
+    bool simplified,
+    MLFloat16* Y_data,
+    U* mean_data,
+    U* inv_std_dev_data) {
+  const MLFloat16* p_input = X_data + task_idx * norm_size;
+  MLFloat16* p_output = Y_data + task_idx * norm_size;
+
+  float mean(0.0f);
+  float mean_square(0.0f);
+
+  const size_t num_elems = static_cast<size_t>(norm_size);
+  float* float_input = new float[num_elems];
+  MlasConvertHalfToFloatBuffer(p_input, float_input, num_elems);
+
+  float* float_output = new float[num_elems];
```

Comment:
>float_output = new float[num_elems [](http://example.com/codeflow?start=9&length=34)

using new inside an op kernel is anti-pattern. Please try using allocator in Compute or prepack.

Reply:
same here